### PR TITLE
CORTX-27237: Collect bytecount data from Motr and store in consul

### PR DESCRIPTION
**Pull request by Shreya-18**
Currently, we have motr apis to fetch the bytecount from motr land.
We need to use these apis and update the byte count of pool versions
in Consul in regular intervals.

Solution:
- Created a separate bytecount updater thread similar to fsStatsUpdater
  which updates bytecount information at regular intervals and stores
  it consul KV.

Signed-off-by: Shreya Karmakar <shreya.karmakar@seagate.com>

### CODE REVIEW

File: hax/hax/bytecount.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            while not self.stopped:
+                if not self.consul._am_i_rc():
+                    wait_for_event(self.event, self.interval_sec)
+                    continue
+                if not motr.is_spiel_ready():
+                    wait_for_event(self.event, self.interval_sec)
+                    continue
+                processes = self.consul.get_proc_fids_with_status(['ios'])
+                for ios, status in processes:
+                    if status == ServiceHealth.OK:
```
Comment (mssawant) [inline hax/hax/bytecount.py:57]: Kindly rebase with latest main, `ServiceHealth` is renamed to `ObjHealth`.

File: hax/hax/bytecount.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        try:
+            LOG.info('byte-count updater thread has started')
+            while not self.stopped:
+                if not self.consul._am_i_rc():
+                    wait_for_event(self.event, self.interval_sec)
+                    continue
+                if not motr.is_spiel_ready():
+                    wait_for_event(self.event, self.interval_sec)
+                    continue
+                processes = self.consul.get_proc_fids_with_status(['ios'])
```
Comment (mssawant) [inline hax/hax/bytecount.py:55]: kindly add a hint of return type for `processes` variable.

File: hax/hax/bytecount.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                for ios, status in processes:
+                    if status == ServiceHealth.OK:
+                        byte_count: ByteCountStats = motr.get_proc_bytecount(
+                            ios)
+                        LOG.debug('Received bytecount: %s', byte_count)
+
```
Comment (mssawant) [inline hax/hax/bytecount.py:61]: [style]: too airy!

File: hax/hax/bytecount.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                        if not byte_count:
+                            continue
+                        try:
+                            self.consul.update_pver_bc(byte_count)
+                        except HAConsistencyException:
+                            LOG.debug('Failed to update Consul KV '
+                                      'due to an intermittent error. The '
+                                      'error is swallowed since new attempts '
+                                      'will be made timely')
+                # testing purpose and usage
```
Comment (mssawant) [inline hax/hax/bytecount.py:71]: ?

File: hax/hax/bytecount.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        LOG.debug('Stop signal received')
+        self.stopped = True
+        self.event.set()
+
+    @log_exception
+    def _execute(self, motr: Motr):
+        try:
+            LOG.info('byte-count updater thread has started')
+            while not self.stopped:
+                if not self.consul._am_i_rc():
```
Comment (mssawant) [inline hax/hax/bytecount.py:49]: s/_am_i_rc/am_i_rc/g
(don't need it to be private now)

File: hax/hax/util.py [hunk 6]
```diff
@@ -487,6 +487,13 @@ def get_session_node(self, session_id: str) -> str:
             raise HAConsistencyException('Failed to communicate to'
                                          ' Consul Agent') from e
 
+    def _am_i_rc(self) -> Any:
```
Comment (mssawant) [inline hax/hax/util.py:490]: s/_am_i_rc/am_i_rc

File: hax/hax/util.py [hunk 7]
```diff
@@ -496,13 +503,15 @@ def get_svc_status(self, srv_fid: Fid, kv_cache=None) -> str:
...
+                         m0d_services=None,
                          kv_cache=None
                          ) -> List[Tuple[ServiceData, ServiceHealth]]:
         """
         Return the list of all Motr service statuses according to Consul
-        watchers. The following services are considered: ios, confd.
+        watchers.The following services are considered [default]: ios, confd.
         """
-        m0d_services = set(['ios', 'confd'])
+        if not m0d_services:
```
Comment (mssawant) [inline hax/hax/util.py:513]: let's rename `m0d_services` to `motr_services`.

File: hax/hax/bytecount.py (hunk 2 above)
Comment (Shreya-18) [reply hax/hax/bytecount.py]: done

File: hax/hax/bytecount.py (hunk 4 above)
Comment (Shreya-18) [reply hax/hax/bytecount.py]: removed .. done

File: hax/hax/bytecount.py (hunk 5 above)
Comment (Shreya-18) [reply hax/hax/bytecount.py]: done

File: hax/hax/util.py (hunk 6 above)
Comment (Shreya-18) [reply hax/hax/util.py]: done

File: hax/hax/util.py (hunk 7 above)
Comment (Shreya-18) [reply hax/hax/util.py]: done

File: hax/hax/bytecount.py (hunk 3 above)
Comment (Shreya-18) [reply hax/hax/bytecount.py]: done

File: hax/hax/bytecount.py (hunk 1 above)
Comment (Shreya-18) [reply hax/hax/bytecount.py]: done

**Comment by Shreya-18** [comment]
Deployment is successful
http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/1140/
```
CORTX Cluster Setup SUCCESS
Please check [cluster setup logs](http://eos-jenkins.colo.seagate.com/job/Cortx-kubernetes/job/deploy-cortx-k8-cluster/1140//console) for more info

Cluster Status:
------------------------------------[ Image Details ]--------------------------------------
ghcr.io/seagate/consul:1.10.0
cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-2169-custom-ci
ghcr.io/seagate/kafka:3.0.0-debian-10-r7
ghcr.io/seagate/symas-openldap:2.4.58
ghcr.io/seagate/zookeeper:3.7.0-debian-10-r182
---------------------------------------[ POD Status ]--------------------------------------
NAME                                                 READY   STATUS    RESTARTS   AGE     IP               NODE                                    NOMINATED NODE   READINESS GATES
consul-client-2lknf                                  1/1     Running   0          10m     172.18.149.196   ssc-vm-g2-rhev4-2610.colo.seagate.com              
consul-client-kcqnx                                  1/1     Running   0          10m     172.18.251.130   ssc-vm-g2-rhev4-2611.colo.seagate.com              
consul-client-vqtvm                                  1/1     Running   0          10m     172.18.189.1     ssc-vm-g2-rhev4-2621.colo.seagate.com              
consul-server-0                                      1/1     Running   0          9m17s   172.18.189.6     ssc-vm-g2-rhev4-2621.colo.seagate.com              
consul-server-1                                      1/1     Running   0          10m     172.18.251.134   ssc-vm-g2-rhev4-2611.colo.seagate.com              
consul-server-2                                      1/1     Running   0          10m     172.18.149.199   ssc-vm-g2-rhev4-2610.colo.seagate.com              
cortx-control-d48f4865f-8hbb2                        4/4     Running   0          5m22s   172.18.189.11    ssc-vm-g2-rhev4-2621.colo.seagate.com              
cortx-data-ssc-vm-g2-rhev4-2610-6dd8b87c68-8hbfl     4/4     Running   0          4m14s   172.18.149.205   ssc-vm-g2-rhev4-2610.colo.seagate.com              
cortx-data-ssc-vm-g2-rhev4-2611-5d97dff4c-b8c4j      4/4     Running   0          4m14s   172.18.251.140   ssc-vm-g2-rhev4-2611.colo.seagate.com              
cortx-data-ssc-vm-g2-rhev4-2621-b4f49df56-zgjx6      4/4     Running   0          4m13s   172.18.189.13    ssc-vm-g2-rhev4-2621.colo.seagate.com              
cortx-ha-679777c9c6-nj26v                            3/3     Running   0          29s     172.18.251.144   ssc-vm-g2-rhev4-2611.colo.seagate.com              
cortx-server-ssc-vm-g2-rhev4-2610-7ccd79f7cd-mjl7t   6/6     Running   0          2m20s   172.18.149.207   ssc-vm-g2-rhev4-2610.colo.seagate.com              
cortx-server-ssc-vm-g2-rhev4-2611-78954f844-qfg67    6/6     Running   0          2m19s   172.18.251.142   ssc-vm-g2-rhev4-2611.colo.seagate.com              
cortx-server-ssc-vm-g2-rhev4-2621-5f6d5c44f4-n5jsr   6/6     Running   0          2m18s   172.18.189.15    ssc-vm-g2-rhev4-2621.colo.seagate.com              
kafka-0                                              1/1     Running   0          7m24s   172.18.189.9     ssc-vm-g2-rhev4-2621.colo.seagate.com              
kafka-1                                              1/1     Running   0          7m24s   172.18.251.138   ssc-vm-g2-rhev4-2611.colo.seagate.com              
kafka-2                                              1/1     Running   0          7m24s   172.18.149.203   ssc-vm-g2-rhev4-2610.colo.seagate.com              
openldap-0                                           1/1     Running   0          10m     172.18.149.198   ssc-vm-g2-rhev4-2610.colo.seagate.com              
openldap-1                                           1/1     Running   0          10m     172.18.251.133   ssc-vm-g2-rhev4-2611.colo.seagate.com              
openldap-2                                           1/1     Running   0          9m54s   172.18.189.4     ssc-vm-g2-rhev4-2621.colo.seagate.com              
zookeeper-0                                          1/1     Running   0          9m18s   172.18.189.7     ssc-vm-g2-rhev4-2621.colo.seagate.com              
zookeeper-1                                          1/1     Running   0          9m18s   172.18.251.136   ssc-vm-g2-rhev4-2611.colo.seagate.com              
zookeeper-2                                          1/1     Running   0          9m18s   172.18.149.201   ssc-vm-g2-rhev4-2610.colo.seagate.com              
-----------[ Sleeping for 1min before checking hctl status.... ]--------------------
---------------------------------------[ hctl status ]-----------------------------------------
Thu Feb 17 03:35:57 MST 2022
Byte_count:
    critical_byte_count : 0
    damaged_byte_count : 0
    degraded_byte_count : 0
    healthy_byte_count : 0
Data pool:
    # fid name
    0x6f00000000000001:0x9c 'storage-set-1__sns'
Profile:
    # fid name: pool(s)
    0x7000000000000001:0xdb 'Profile_the_pool': 'storage-set-1__sns' 'storage-set-1__dix' None
Services:
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2621 
    [started]  hax        0x7200000000000001:0x2b  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@22001
    [started]  ioservice  0x7200000000000001:0x2e  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21001
    [started]  ioservice  0x7200000000000001:0x3b  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21002
    [started]  confd      0x7200000000000001:0x48  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2621@21003
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2610  (RC)
    [started]  hax        0x7200000000000001:0x7   inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@22001
    [started]  ioservice  0x7200000000000001:0xa   inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21001
    [started]  ioservice  0x7200000000000001:0x17  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21002
    [started]  confd      0x7200000000000001:0x24  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2610@21003
    cortx-data-headless-svc-ssc-vm-g2-rhev4-2611 
    [started]  hax        0x7200000000000001:0x4f  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@22001
    [started]  ioservice  0x7200000000000001:0x52  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21001
    [started]  ioservice  0x7200000000000001:0x5f  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21002
    [started]  confd      0x7200000000000001:0x6c  inet:tcp:cortx-data-headless-svc-ssc-vm-g2-rhev4-2611@21003
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2610 
    [started]  hax        0x7200000000000001:0x71  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2610@22001
    [started]  s3server   0x7200000000000001:0x74  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2610@22501
    [started]  s3server   0x7200000000000001:0x77  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2610@22502
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2611 
    [started]  hax        0x7200000000000001:0x7c  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2611@22001
    [started]  s3server   0x7200000000000001:0x7f  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2611@22501
    [started]  s3server   0x7200000000000001:0x82  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2611@22502
    cortx-server-headless-svc-ssc-vm-g2-rhev4-2621 
    [started]  hax        0x7200000000000001:0x87  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2621@22001
    [started]  s3server   0x7200000000000001:0x8a  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2621@22501
    [started]  s3server   0x7200000000000001:0x8d  inet:tcp:cortx-server-headless-svc-ssc-vm-g2-rhev4-2621@22502
-----------[ Time taken for service to start 0 mins ]--------------------
```

consul kv is updated correctly!
```
[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-2610 /]# consul kv get -recurse ios
ioservices/0x7200000000000001:0x17/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
ioservices/0x7200000000000001:0x2e/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
ioservices/0x7200000000000001:0x3b/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
ioservices/0x7200000000000001:0x52/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
ioservices/0x7200000000000001:0x5f/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
ioservices/0x7200000000000001:0xa/pvers/0x7600000000000001:0x8/users/8881212:{"bc": 10240000, "object_cnt": 10000}
[root@cortx-data-headless-svc-ssc-vm-g2-rhev4-2610 /]#
```

**Review by mssawant — APPROVED** [review]

**Comment by vaibhavparatwar** [comment]
@mssawant @suvratjoshi - can we merge it to feature branch now?

**Pull request merged**